docs/architecture/adr-051-arbitrary-protobuf-ipld-support-scheme.md

```diff
@@ -0,0 +1,228 @@
+# ADR-051: Arbitrary Protobuf IPLD Support Scheme
+
+## Changelog
+
+- Feb 14th, 2022: Initial Draft
+
+## Status
+
+DRAFT
+
+## Abstract
+This ADR describes a generic IPLD content addressing scheme for the arbitrary protobuf types stored in SMT referenced state storage.
+
+## Context
+
+Because all values stored in the state of a ComsosSDK chain are protobuf encoded, we are presented with a unique
+opportunity to provide extremely rich IPLD content-typing information for the arbitrary types stored in
+(referenced from) the SMT in a generic fashion.
+
+### SDK context
+
+The SDK stores values in state storage in a protobuf encoded format. Each module defines their own custom types in
+.proto definitions and these types are registered with a ProtoCodec that manages the marshalling and unmarshalling of these
+types to and from the binary format that is persisted to disk. As such, an indefinite/unbounded number of protobuf types
+need to be supported within the Cosmos ecosystem at large.
+
+Rather than needing to register new content types and implement custom codecs for every Cosmos protobuf type we
```

Comment:
Hey @marbar3778 , closed this PR as requested but if you get a chance to read through my response to this comment (https://github.com/cosmos/cosmos-sdk/pull/11186#issuecomment-1068448614) let me know what you think. If I am mistaken and I missed some grpc reflection feature that can fully substitute for what is proposed here it would be great to know! Thanks!